sw/device/lib/dif/README.md

```diff
@@ -48,9 +58,47 @@ there are some relaxations of these rules for them described at the end.
     in DIF signatures.
 
 * DIFs must use enum return codes rather than booleans for reporting errors. If
-  a DIF can either error or instead produce a value, it should return an enum
+  a DIF can either error or instead produce a value, it must return an enum
   return code, and use an out-parameter for returning the produced value.
-  * DIFs that cannot error and that do not return a value should return `void`.
+  * DIF libraries must have a general return code enum called
+    `dif_<ip>_result_t`. It must define the following constants (in this order):
+    * `kDif<ip>Ok`, with value 0, to denote the call succeeded.
+    * `kDif<ip>Error`, with value 1, to denote a non-specific error happened
+      during the call. This is for the `default:` case of enum switches (as
+      noted below), and for assertion errors (usually where the function has
+      already caused side-effects so `kDif<ip>BadArg` cannot be used).
+    * `kDif<ip>BadArg`, with value 2, to denote that the caller supplied
+      incorrect arguments. This value must only be returned if the function has
+      not caused any side-effects.
+  * DIF libraries should define specific return code enums for operations that
+    fail in more specific ways. These specific return code enums can be shared
+    between multiple DIFs that fail in the same way.
+
+    These more specific return code types must be named
+    `dif_<ip>_<operation>_result_t`, and their constants must be prefixed with
+    `kDif<ip><operation>`. `<operation>` need not correspond to a DIF name if
+    the return codes are shared between DIFs.
+
+    The first three constants in these specific enums must define the following
+    constants:
+    * `kDif<ip><operation>Ok`, with value `kDif<ip>Ok`,
+    * `kDif<ip><operation>Error`, with value `kDif<ip>Error`, and
+    * `kDif<ip><operation>BadArg`, with value `kDif<ip>BadArg`.
+
+    Additional, specific return code constants must all be defined after
+    these three general constants, and may cover more specific forms of the
+    return codes defined above, including more specific reasons arguments are
+    invalid.
+
+  * DIFs must document the meaning of all return codes, including the required
```

Comment:
Nit: I assume this means that the return code enum must contain doxygen comments for all of its variants; maybe we should clarify this?

(I could imagine someone thinking that it would need to be documented on the DIF's `@return` comment instead...)